File: gateway/src/main/kotlin/com/qianlei/gateway/service/Service.kt
package com.qianlei.gateway.service

import com.qianlei.gateway.config.Node
import com.qianlei.gateway.constant.LoadBalanceType
import com.qianlei.gateway.constant.LoadBalanceType.*
import com.qianlei.gateway.service.loadbalance.HashLoadBalance
import com.qianlei.gateway.service.loadbalance.ILoadBalance
import com.qianlei.gateway.service.loadbalance.RandomLoadBalance
import com.qianlei.gateway.service.loadbalance.RoundRibbonLoadBalance
import kotlinx.serialization.Serializable

/**
 *
 * @author qianlei
 */
@Serializable
data class Service(
    val type: LoadBalanceType = ROUND_RIBBON,
    val nodes: List<Node> = emptyList(),
    val hashOn: String = ""
) {
    fun loadBalance() = createLoadBalance()

    private fun createLoadBalance(): ILoadBalance {
        return when (type) {
            ROUND_RIBBON -> RoundRibbonLoadBalance(nodes)
            HASH -> HashLoadBalance(nodes, hashOn)
            RANDOM -> RandomLoadBalance(nodes)
        }
    }
}

File: gateway/service/service.go
package service

import (
	"encoding/json"
	"gateway/service/loadbalance"
	"gateway/upstream"
)

type Service struct {
	LoadBalance loadbalance.LoadBalance
	HashOn      string
	Nodes       []upstream.Node
}

const (
	RANDOM      = "random"
	RoundRibbon = "round_ribbon"
	HASH        = "hash"
)

func (s *Service) UnmarshalJSON(data []byte) error {
	ss := struct {
		LoadBalanceType string `json:"load_balance_type"`
		HashOn          string
		Nodes           []upstream.Node `json:"nodes"`
	}{}

	if err := json.Unmarshal(data, &ss); err != nil {
		return err
	}
	s.Nodes = ss.Nodes
	switch ss.LoadBalanceType {
	case RANDOM:
		s.LoadBalance = &loadbalance.RandomLoadBalance{}
	case RoundRibbon:
		s.LoadBalance = &loadbalance.RoundRibbonLoadBalance{}
	case HASH:
		s.LoadBalance = &loadbalance.HashLoadBalance{
			HashOn: ss.HashOn,
		}
	}
	return nil
}

func (s Service) MarshalJSON() ([]byte, error) {
	var lbtype = ""
	var hashOn = ""
	switch s.LoadBalance.(type) {
	case *loadbalance.RandomLoadBalance:
		lbtype = RANDOM
	case *loadbalance.RoundRibbonLoadBalance:
		lbtype = RoundRibbon
	case *loadbalance.HashLoadBalance:
		lbtype = HASH
		hashOn = s.LoadBalance.(*loadbalance.HashLoadBalance).HashOn
	}
	return json.Marshal(struct {
		LoadBalanceType string `json:"load_balance_type"`
		HashOn          string
		Nodes           []upstream.Node `json:"nodes"`
	}{
		LoadBalanceType: lbtype,
		HashOn:          hashOn,
		Nodes:           s.Nodes,
	})
}
